src/main/java/net/minestom/server/event/Event.java

```diff
@@ -1,11 +1,94 @@
 package net.minestom.server.event;
 
 import net.minestom.server.event.handler.EventHandler;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.function.Consumer;
+import java.util.function.Function;
 
 /**
  * Object which can be listened to by an {@link EventHandler}.
  * <p>
  * Called using {@link EventHandler#callEvent(Class, Event)}.
  */
 public class Event {
+
+    public static <T extends EntityEvent> Listener.Builder<T> entity(Class<T> eventType) {
+        return new Listener.Builder<>(eventType);
+    }
+
+    public static <T extends PlayerEvent> Listener.Builder<T> player(Class<T> eventType) {
+        return new Listener.Builder<>(eventType);
+    }
+
+
+    public static class Listener<T extends Event> {
+
+        private final Class<T> eventType;
+        private final Set<EventHandler> attach = new CopyOnWriteArraySet<>();
+        private final Consumer<T> combined;
+
+        private Listener(Class<T> eventType, @NotNull Consumer<T> combined) {
+            this.eventType = eventType;
+            this.combined = combined;
+        }
+
+        public void attachTo(@NotNull EventHandler handler) {
```

Comment:
I'm not so sure about this. Why do we need to keep a set of the handlers?

Reply:
So we can have a method to detach it from all the handlers. Would be really powerful once listeners group are implemented